ADR/00xx-component-dependencies.md

```diff
@@ -0,0 +1,171 @@
+* [integration-service] will always skip testing for parent image update builds, will never promote them, or use them to initiate Releases, but it will promote them to the global candidate list.
+* [build-service] will propagate digest references as a PR to the child image Component repo, by analyzing the `depends on` fields of all other Components in the Application. The PR that it files must be submitted in such a way that it appears in the same PR Group as the triggering PR submitted by a user.
+* When the parent image PR is merged, [build-service] will update the PRs it originally filed to take them out of "Draft" to indicate that they are safe to merge now as long as there are no other unmerged triggering PRs in the same group. It may potentially rebase the PRs to trigger a new build or use `/retest`.
+
+Think about branches:
+
+* User team updates parent image with branch `update-2023-06-07` (or whatever).
+* [build-service] uses that branch name as a prefix in the branch names that it chooses (`update-2023-06-07/<suffix>`) everywhere it propagates the digest to.
+
+### OLM Operators, with components in different repos
+
+Scenario: the user has 5 components. One of them is a “bundle” image. It contains references to the other four images by digest. None of the images are in the same repo.
+
+* The bundle Component declares that it `depends on` all other operand image Components, by way of the new field on the Component CR.
+* [integration-service] will always skip testing for all components except for the bundle image Component, because the bundle image Component declares that it depends on all other Components. (All other Components are **dependency** Components. The bundle is the only **dependant** Component.)
+* [build-service] will propagate digest references as a PR to the bundle image, in the same PR Group
+  as the originating PR. The PR will be marked as "Draft" to indicate they should not be merged.
+* When the operand image PR is merged, [build-service] will update the PRs it originally filed to
+  take them out of "Draft" to indicate that they are safe to merge now as long as there are no other
+  unmerged triggering PRs in the same group.
+
+Think about branches:
+
+* User team updates their operand images with branch `feature-1234` (or whatever).
+* [build-service] uses that same branch in its PR to the bundle repo.
+* [build-service] uses that branch name as a prefix in the branch names that it chooses (`feature-1234/<suffix>`) to send to the bundle repo.
+* If user submits three operand image PRs and uses the same branch name for all (`feature-1234`) – then the digests all pile up on the same bundle branch (`feature-1234/<suffix>`), the same bundle PR.
+* If the user submits three operand image PRs and uses different branch names for all – then the digests are split among three different bundle branches, three different bundle PRs.
+
+---
+
+### OLM Operators, with components in the same repo
```

Comment:
Yeah, I see. `ko` looks really straightforward; `hack/release.sh` is super simple. (I'm having a hard time imagining how `ko` could be used in RHTAP pipelines, since we have a 1 to 1 relationship between components and pipelines and between components and images.)

Trying to play it out with the proposal here:

* The user has the repo modelled as ~6 Components. Four for the single-binary images, one for the controller, and one for release.yaml. Because RHTAP is container centric, the user has to wrap release.yaml with a dockerfile. That 6th RHTAP component is just a container that contains a built release.yaml. The user specifies that the release.yaml Component depends on all other 5 Components.
* An important detail is that the controller does not embed references to the other images. The reference only exists in release.yaml, in the environment variables.
* A user would propose a change in one PR to the controller and to one of the single-binary images.
* In RHTAP, this would trigger two PipelineRuns. One for the controller and one for one of the single-binary images.
* In RHTAP, build-service would respond to whichever PLR finished first (let's say its the controller) and submit a followup PR in draft mode to the repo, updating the digest reference for the controller.
* In RHTAP, build-service would respond to the second PLR finishing (the single-binary image) and *amends* the PR it just submitted by adding another commit updating the digest reference for the single-binary image. That draft PR now includes both changes. build-service knows to group these two updates into the same PR, because the originating changes both came from the same branch -- they are from the same "PR Group".

Note -- the above assumes that pipelines-as-code has been modified to somehow not start pipelineruns for the components which are not having their source modified in the originating PR. It doesn't do this today (see also [RHTAP-371](https://issues.redhat.com/browse/RHTAP-371).

Reply:
I guess the interesting thing happens when you want to make a co-dependent change in any one of the image _and in the files in `deploy/` (i.e. release.yaml) at the same time. Let me try again.

Same setup as before, except this time:

* A user proposes a change in one PR to the controller and to `deploy/500-controller.yaml` at the same time.
* In RHTAP, this would trigger two PLRs, one for the controller and one for release.yaml.
* In RHTAP, build-service would respond to whichever PLR finished first (let's say its the controller) and submit a followup PR in draft mode to the repo, updating the digest reference for the controller.
* In RHTAP, build-service would respond to the second PLR finishing (this is the one for release.yaml). But -- here, it would do nothing, because it knows that no other Components depend on release.yaml. This image (since there are no dependencies on it) would be tested independently and would probably fail, since we assume the user is making meaningful changes to it at the same time. Meanwhile, the test of the controller change probably also fails because it can't get itself associated with a build of release.yaml that contains both the digest update and whatever the meaningful change to `deploy/500-controller.yaml` was.

Reply:
OK, with two more changes (5d1b9ff first and 64f37c5 second), this is more workable although - it's awkward.

Setup: the user has 6 Components as before. Each one has its `context` dir set to something reasonable. The release component's context dir is set to `deploy/` and the controller's context dir is set to `cmd/shipwright-build-controller/` for instance (see 5d1b9ff).

* A user proposes a change in one PR to the controller and to `deploy/500-controller.yaml at the same time.
* In RHTAP, this would trigger two PLRs, one for the controller and one for release.yaml.
* In RHTAP, build-service responds to the release.yaml PLR finishing for PR#1. It builds. It initiates testing. Testing fails because it is not in a Snapshot that also contains the controller changes.
* In RHTAP, build-service responds to the controller PLR finishing for PR#1. It builds. Due to declared `depends on` links in the Component, integration-service skips testing. build-service submits a followup PR (PR#2) in draft mode back to the repo, updating the digest reference for the controller.
  * PR#2 contains not only the digest reference update in `deploy/500-controller.yaml`, but also the changes introduced by the user to the `deploy/` directory in PR#1 (see 64f37c5).
* In RHTAP, PaC receives notice of PR#2 and kicks of a single Component build of the release component. There are no changes in PR#2 beyond the `deploy/` directory.
* build-service propagates no new PLRs in response to this release PLR finishing for PR#2. There are no other Components which depend on it.
* integration-service responds to this release PLR finishing for PR#2. It creates a snapshot composed of the image of the release component from the PLR in PR#2 as well as the controller component from the PLR in PR#1. It tests these together. It passes. integration-service reports success back to both PR#1 and PR#2.

Reply:
This hinges on the ability for AppStudio to build components separately just based on their context directory being changed. Is that in scope for this ADR or is it a required dependency of it? It isn't something that is currently supported.

I think that this same situation is easily replicated by many operands as the code for the bundle images are often (i.e. by default) co-located in the same git repo as the operator code. An operator would be `depends-on` relationship with the bundle, but when the bundle change is submitted (i.e. updating the reference to the operator), then the operator itself will rebuild. This will result in an inability to achieve eventual snapshot consistency as the bundle will never be able to be built properly referencing the operator.

Instead of trying to fit this model into the component reference, would it make sense to enable a process by which a reference isn't image digest pinned, but instead it is git commit pinned? This change would require AppStudio to properly schedule the pipelines if there are multiple dependent components in the same git repo, but then any push to the repo would result in consistency between all components.

In the case above, the `depends-on` relationship would need to be made for all components so that they are not built in parallel.

In this situation, when the _primary_ PR is merged in the git repo, then instead of marking the other PRs as mergable, it would instead close them. All components would be rebuilt as desired from a single commit.

Reply:
> Is that in scope for this ADR or is it a required dependency of it? It isn't something that is currently supported.

Yeah, I've been thinking of it as a pre-requisite.

> Instead of trying to fit this model into the component reference, would it make sense to enable a process by which a reference isn't image digest pinned, but instead it is git commit pinned?

Yeah, the desired behavior that Luiz raised for the contract team reminds me of this same flow.

Reply:
If nothing else, we should state that there is an external dependency here. The resolution of this external dependency can come later as it is specific to the use case of components colocated in a git repository. We can explore which of these two options might be better received by users before determining which to go with.

Reply:
Added in 9486e34bce2f4116b58ea5875bf5786c7417a5e7.